### PR TITLE
Migrate usage of EC2 to aws-sdk-go-v2

### LINKED ISSUE #4883: migrate to aws-sdk-go-v2 [7/9]
## Background

As part of #3215 we want to migrate to use the [aws-sdk-go v2](https://github.com/aws/aws-sdk-go-v2). This involves migrating all of the AWS API calls, roughly 500~ to use the new package. A spike was complete and its finding was documented [here](https://github.com/weaveworks/eksctl/issues/3215#issuecomment-1047099306) and that included a demo branch [here](https://github.com/weaveworks/eksctl/pull/4820/files). As the spike documents, we can split out the work to migrate to v2 to be done in 9 parts, where we essentially use both v1 and v2 of the API while we migrate.

This ticket is to: remove usage of `v1` API for `EC2` and replace it with `v2`. Consider doing in multiple PRs (where EC2 and EC2 V2 can live alongside each other)

Blocked on #4877. Can be done in parallel to parts 2-8
Also, when migrating to a new call we will propagate the context from the command down to the AWS calls.

### PR DESCRIPTION
**Pull request by cPu1**
### Description

This changelist migrates usage of EC2 to aws-sdk-go-v2. The integration tests are still using aws-sdk-go for EC2 and other services, it's not addressed in this PR as the setup code for SDK v2 is missing in integration tests. I will open a separate ticket for migrating usage of AWS services in integration tests. 

This branch is based off `migrate-iam-sdk-v2`.

Closes #4883 

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [ ] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/az/az.go [hunk 1]
```diff
@@ -1,13 +1,18 @@
 package az
 
 import (
+	"context"
 	"fmt"
 	"math/rand"
 	"time"
 
+	ec22 "github.com/aws/aws-sdk-go-v2/service/ec2"
```
Comment (aclevername) [inline pkg/az/az.go:9]: :eyes:

Comment (cPu1) [reply pkg/az/az.go:9]: There are some other files that have this too, I'm fixing it. (I have had to rebase and resolve conflicts way too many times).

**Comment by Skarlso** [comment]
👀 👀 👀 
```
+20,649 −46,976
```

:D lots of generated lines

**Comment by cPu1** [comment]
Please note that I have a couple of TODO comments, mainly around testing error types and codes. I'll remove them after testing.

File: pkg/cfn/builder/managed_nodegroup.go [hunk 2]
```diff
@@ -149,18 +152,19 @@ func (m *ManagedNodeGroupResourceSet) AddAllResources() error {
 		}
 
 		if launchTemplateData.ImageId == nil {
-			if launchTemplateData.InstanceType == nil {
+			// TODO: what?
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply pkg/cfn/builder/managed_nodegroup.go:155]: Sad that there is no troll react emoji. :D

[the remark this answers is not in the document]
Comment (cPu1) [reply pkg/cfn/builder/managed_nodegroup.go:155]: haha, this is one of the [TODOs](https://github.com/weaveworks/eksctl/pull/5086#issuecomment-1091702334) that I need to test. `InstanceType` was a pointer before but in v2 it's a value type, so I need to test that it's being set to the empty string when it's unset.

File: pkg/cfn/builder/network_interfaces.go [hunk 3]
```diff
@@ -22,29 +26,34 @@ func defaultNetworkInterface(securityGroups []*gfnt.Value, device, card int) gfn
...
 			return errors.Wrapf(err, "couldn't retrieve instance type description for %v", instanceTypes)
 		}
 
 		var numEFAs = math.MaxFloat64
 		for _, it := range info.InstanceTypes {
 			networkInfo := it.NetworkInfo
-			numEFAs = math.Min(float64(aws.Int64Value(networkInfo.MaximumNetworkCards)), numEFAs)
-			if !aws.BoolValue(networkInfo.EfaSupported) {
-				return errors.Errorf("instance type %s does not support EFA", *it.InstanceType)
+			numEFAs = math.Min(float64(aws.ToInt32(networkInfo.MaximumNetworkCards)), numEFAs)
```
[the remark this answers is not in the document]
Comment (cPu1) [reply pkg/cfn/builder/network_interfaces.go:54]: This is using the v2 package `github.com/aws/aws-sdk-go-v2/aws`.

File: pkg/kops/kops.go [hunk 4]
```diff
@@ -41,33 +43,50 @@ func (k *Wrapper) topologyOf(s *ec2.Subnet) api.SubnetTopology {
...
+func (k *Wrapper) UseVPC(ctx context.Context, ec2API awsapi.EC2, spec *api.ClusterConfig) error {
 	spec.VPC.CIDR = nil // ensure to reset the CIDR
 
-	allSubnets, err := aws.ListSubnets(k.cloud, k.clusterName)
+	clusterTag := fmt.Sprintf("kubernetes.io/cluster/%s", k.clusterName)
+	output, err := ec2API.DescribeSubnets(ctx, &ec2.DescribeSubnetsInput{
+		Filters: []ec2types.Filter{
+			{
+				Name:   aws.String(fmt.Sprintf("tag:%s", clusterTag)),
+				Values: []string{k.clusterName},
```
Comment (Skarlso) [inline pkg/kops/kops.go:54]: Is this correct? Previously it was checking if the value is "owned". Now it's checking if the value is the clusterName. Did that change?

**Review by Skarlso — APPROVED** [review]
One question / uncertainty otherwise LGTM. Phew that was a large change, but mostly just renaming and context passing.

File: pkg/kops/kops.go (hunk 4 above)
Comment (cPu1) [reply pkg/kops/kops.go:54]: Good catch. The value should be `owned` here.

Comment (Skarlso) [reply pkg/kops/kops.go]: 🎉

**Review by Himangini — APPROVED** [review]
LGTM 👍🏻

**Comment by cPu1** [comment]
> Please note that I have a couple of TODO comments, mainly around testing error types and codes. I'll remove them after testing.

I have tested all the TODOs and removed the comments, in addition to resolving the conflicts.
